### DISCUSSION TITLE
Passing dots arguments through a target factory

### DISCUSSION THREAD

**Discussion by petrbouchal** [ropensci/tarchetypes#84] [Help] [answered]
I am trying to build a target factory which is a combination of a file target and a target running a read function - it should take a path and a function as arguments, and return two targets: a file target and a the result of reading `file` using `function`. 

e.g.

```r
tar_read_data(data, "data.csv", read_csv, skip = 5)
```

where `skip` is passed via `...`. But I am completely stumped with the metaprogramming of passing `...` into the expression for the second target. Is there a function in tarchetypes or elsewhere that I could learn from? I looked at `tar_render()` but was not able to make sense of how it works with `...`.

More generally, something along these lines might be a good addition in tarchetypes? I feel I use the pattern of these two targets quite often.

**Comment by wlandau** [comment, marked as answer]
`rlang` has nice functions for this, especially `enexprs()`. `rlang::call2()` almost fits, but `base::as.call()` may be a little better for the situation. 

``` r
library(targets)

tar_script({
  tar_read_data <- function(name, file, reader, ...) {
    name_read <- deparse(rlang::enexpr(name))
    name_file <- paste0(name_read, "_file")
    sym_file <- rlang::sym(name_file)
    expr_read <- as.call(rlang::enexprs(reader, sym_file, ...))
    list(
      targets::tar_target_raw(name_file, file, format = "file"),
      targets::tar_target_raw(name_read, expr_read)
    )
  }

  tar_read_data(data, "data.csv", read_csv, skip = 5, col_types = cols())
})

tar_manifest(fields = all_of(c("command", "format")))
#> # A tibble: 2 × 3
#>   name      command                                             format
#>   <chr>     <chr>                                               <chr> 
#> 1 data_file "\"data.csv\""                                      file  
#> 2 data      "read_csv(data_file, skip = 5, col_types = cols())" rds
```

<sup>Created on 2022-01-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Could be a nice addition to `tarchetypes`, with a clearer name and some refinement to accept other `tar_target()` arguments like `storage` and `memory`.

**Reply by petrbouchal** [reply]
Excellent, thank you! 

For tarchetypes, a nice complement to this would be the analogue write factory, which would just return a single file target, but unlike tar_file would also take a fun and a … argument - but the gain in brevity would be less in this case.

**Reply by petrbouchal** [reply]
Which I now realise would be just a wrapper

**Reply by wlandau** [reply]
I just implemented `tar_file_read()`. Let's see how that goes for now.

**Reply by petrbouchal** [reply]
Thank you, that looks great.

Very much appreciate your responsiveness!
